### PR TITLE
NEW: Save and load rebinds as JSON.

### PR DESCRIPTION
**Pull request by Rene-Damm**
This started out as "just some docs improvement/fixes" but then ended up accumulating several improvements to InputAction-related APIs for things that otherwise took a lot of text in docs to explain how to do. Nothing you couldn't do before but all stuff that before took a lot of knowing intricacies.

Changes:
* A bunch of .md updates.
* Added APIs to save and load action rebinds in JSON format. Hotly requested feature.
* Added APIs to make it easier to navigate bindings and alter them.

### CODE REVIEW

File: CONTRIBUTIONS.md [hunk 1]
```diff
@@ -0,0 +1,21 @@
+# Contributions
+
+## If you are interested in contributing, here are some ground rules:
+* Ideally, [talk to us](https://forum.unity.com/forums/input-system.103/) before doing the work -- we appreciate contributions, but we might already be working on the same thing, or we might have different opinions on how it should be implemented.
+
+## All contributions are subject to the [Unity Contribution Agreement(UCA)](https://unity3d.com/legal/licenses/Unity_Contribution_Agreement)
```
Comment (karljj1) [inline CONTRIBUTIONS.md:6]: Does this come from Legal? Its worth getting them to check and make sure the wording is correct.

File: Packages/com.unity.inputsystem/Documentation~/Testing.md [hunk 2]
```diff
@@ -7,7 +7,7 @@ The Input System has built-in support for writing automated input tests. You can
 To set up a test assembly that uses the Input System's automation framework, follow these steps:
 
 1. In the `Packages/manifest.json` file of your project, `com.unity.inputsystem` must be listed in `testables`. This is necessary for test code that comes with the package to be included with test builds of your project.<br><br>You can, for example, add this after the `dependencies` property like so:
-    ```JSON
```
Comment (karljj1) [inline Packages/com.unity.inputsystem/Documentation~/Testing.md:10]: Why did you remove the json tag?

File: Packages/com.unity.inputsystem/InputSystem/Actions/IInputActionCollection.cs [hunk 3]
```diff
@@ -56,11 +57,78 @@ public interface IInputActionCollection : IEnumerable<InputAction>
...
         void Disable();
     }
+
+    /// <summary>
+    /// An extended version of <see cref="IInputActionCollection"/>.
+    /// </summary>
+    /// <remarks>
+    /// This interface will be merged into <see cref="IInputActionCollection"/> in a future (major) version.
+    /// </remarks>
+    public interface IInputActionCollection2 : IInputActionCollection
```
Comment (karljj1) [inline Packages/com.unity.inputsystem/InputSystem/Actions/IInputActionCollection.cs:78]: Perhaps **IInputActionCollectionExtended** would be a better name?
Could some of this be done through extensions so that you can already support IInputActionCollection?

**Review by karljj1 — APPROVED** [review]

File: Packages/com.unity.inputsystem/Documentation~/Testing.md (hunk 2 above)
Comment (Rene-Damm) [reply Packages/com.unity.inputsystem/Documentation~/Testing.md:10]: JSON doesn't seem to be a recognized language so the code block ends up garbled (see https://docs.unity3d.com/Packages/com.unity.inputsystem@1.0/manual/Testing.html).

File: Packages/com.unity.inputsystem/InputSystem/Actions/IInputActionCollection.cs (hunk 3 above)
Comment (Rene-Damm) [reply Packages/com.unity.inputsystem/InputSystem/Actions/IInputActionCollection.cs:78]: `IInputActionCollectionExtended` IMO makes it sound like it's rightfully a different thing from `IInputActionCollection`. Tagging a number on it seemed like the clearest way to mark it as an artifact of versioning.

Comment (Rene-Damm) [reply Packages/com.unity.inputsystem/InputSystem/Actions/IInputActionCollection.cs:78]: Ah yup, did consider extension methods but unfortunately, at least for some of the stuff here, wasn't an option (e.g. `bindings` can't be implemented properly with just the stuff in `IInputActionCollection`).

File: CONTRIBUTIONS.md (hunk 1 above)
Comment (Rene-Damm) [reply CONTRIBUTIONS.md:6]: This one is copied from the various other CONTRIBUTIONS.md we have. Probably a bit cheap on my end but with some of that stuff already on trunk and with the doc itself up on unity3d.com/legal, I just told myself "this is already properly vetted"  :)

File: CONTRIBUTIONS.md [hunk 4]
```diff
@@ -0,0 +1,21 @@
+# Contributions
+
+## If you are interested in contributing, here are some ground rules:
+* Ideally, [talk to us](https://forum.unity.com/forums/input-system.103/) before doing the work -- we appreciate contributions, but we might already be working on the same thing, or we might have different opinions on how it should be implemented.
+
+## All contributions are subject to the [Unity Contribution Agreement(UCA)](https://unity3d.com/legal/licenses/Unity_Contribution_Agreement)
+By making a pull request, you are confirming agreement to the terms and conditions of the UCA, including that your Contributions are your original creation and that you have complete right and authority to make your Contributions.
+
+## Once you have a change ready following these ground rules. Simply make a pull request in Github
```
Comment (keiramoon) [inline CONTRIBUTIONS.md]:
```suggestion
## Once you have a change ready that complies to the rules above, open a pull request in GitHub.

```

File: CONTRIBUTIONS.md [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+## If you are interested in contributing, here are some ground rules:
+* Ideally, [talk to us](https://forum.unity.com/forums/input-system.103/) before doing the work -- we appreciate contributions, but we might already be working on the same thing, or we might have different opinions on how it should be implemented.
+
+## All contributions are subject to the [Unity Contribution Agreement(UCA)](https://unity3d.com/legal/licenses/Unity_Contribution_Agreement)
+By making a pull request, you are confirming agreement to the terms and conditions of the UCA, including that your Contributions are your original creation and that you have complete right and authority to make your Contributions.
+
+## Once you have a change ready following these ground rules. Simply make a pull request in Github
+
+Some notes:
```
Comment (keiramoon) [inline CONTRIBUTIONS.md]:
```suggestion
Notes:

```

File: CONTRIBUTIONS.md [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+
+* We can rarely take a pull request directly and as is. The reason for this is that our CI will not be run on PRs coming from forks. So, in most cases we will have to incorporate your changes into a branch inside the repo and PR it into the `develop` branch via a separate PR. When we do so, we will clearly state that in your PR.
```
Comment (keiramoon) [inline CONTRIBUTIONS.md]:
```suggestion
* We can rarely take a pull request directly and as-is because our CI doesn't run on PRs coming from forks. So, in most cases we will have to incorporate your changes into a branch inside the repository and merge it into the `develop` branch via a separate PR. When we do so, we will clearly state that in your PR.

```

Comment (keiramoon) [reply CONTRIBUTIONS.md]: Do we need to tell them why we can't take PRs directly? The bit about our CI not running on PRs coming from forks seems superfluous to me (but that might just be my minimalist brain at work).

File: CONTRIBUTIONS.md [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+* All changes are subject to the following requirements:
+  1) Tests need to be green.
+  2) Behavioral changes need to be accompanied by test changes/additions.
+  3) Where relevant, documentation (xmldoc and/or manual) needs to be updated.
```
Comment (keiramoon) [inline CONTRIBUTIONS.md]:
```suggestion
  3) Where relevant, documentation (xmldoc and/or Manual) needs to be updated.

```

File: CONTRIBUTIONS.md [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+  4) Fixes, changes, and additions need to be documented in [CHANGELOG](./CHANGELOG.md).
+  5) Breaking API changes can only be introduced in new major versions.
+* The more a PR deviates from these requirements, the more work we have to do on our side to incorporate your changes. Give your PR a leg up by making it easy on us :)
+* We are hoping to have better/easier flows for incorporating user PRs in the future.
```
Comment (keiramoon) [inline CONTRIBUTIONS.md]:
```suggestion
* We are hoping to have an easier workflow for incorporating user PRs in the future.
```

Comment (keiramoon) [reply CONTRIBUTIONS.md]: Just a style edit to avoid the slash.

File: CONTRIBUTIONS.md [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+By making a pull request, you are confirming agreement to the terms and conditions of the UCA, including that your Contributions are your original creation and that you have complete right and authority to make your Contributions.
+
+## Once you have a change ready following these ground rules. Simply make a pull request in Github
+
+Some notes:
+
+* We can rarely take a pull request directly and as is. The reason for this is that our CI will not be run on PRs coming from forks. So, in most cases we will have to incorporate your changes into a branch inside the repo and PR it into the `develop` branch via a separate PR. When we do so, we will clearly state that in your PR.
+* All changes are subject to the following requirements:
+  1) Tests need to be green.
+  2) Behavioral changes need to be accompanied by test changes/additions.
```
Comment (keiramoon) [inline CONTRIBUTIONS.md:16]:
```suggestion
  2) Behavioral changes need to be accompanied by test changes and additions.

```

Comment (keiramoon) [reply CONTRIBUTIONS.md:16]: Do we mean that tests need to cover any behavioral changes? If so, we could probably rephrase this to make it clearer.

File: Packages/com.unity.inputsystem/CHANGELOG.md [hunk 10]
```diff
@@ -49,6 +50,23 @@ however, it has to be formatted properly to pass verification tests.
 - Device layouts can now be "precompiled" for speed. `Keyboard`, `Mouse`, and `Touchscreen` are now included as precompiled layouts greatly reducing instantiation time and GC heap cost for these devices. For `Touchscreen`, this results in a >20x speed-up for `InputSystem.AddDevice<Touchscreen>()`.
 - Added Pose Control layout. The Pose Control is used on XR Devices and wraps tracking state, position, rotation, and velocity information.
 
+#### Actions
+
+- Can now save binding overrides as JSON strings and restore them from such using the newly added `SaveBindingOverridesAsJson` and `LoadBindingOverridesFromJson` extension methods.
```
Comment (keiramoon) [inline Packages/com.unity.inputsystem/CHANGELOG.md:55]:
```suggestion
- You can now save binding overrides as JSON strings and restore them from these strings using the newly added `SaveBindingOverridesAsJson` and `LoadBindingOverridesFromJson` extension methods.
```

File: Packages/com.unity.inputsystem/Documentation~/ActionBindings.md [hunk 11]
```diff
@@ -261,58 +286,162 @@ public class CustomParameterEditor : InputParameterEditor<CustomComposite>
...
-Actions with the type set to [Button](Actions.md#button) don't perform any initial state check, so that only buttons pressed after the Action was enabled have any effect on the Action.
+//
+playerInput.actions["move"].ChangeBinding("WASD")
+```
 
-## Runtime rebinding
+### Applying overrides
 
-Runtime rebinding allows users of your application to set their own Bindings.
+You can override aspects of any Binding at run-time non-destructively. Specific properties of [`InputBinding`](../api/UnityEngine.InputSystem.InputBinding.html) have an `override` variant that, if set, will take precedent over the property that they shadow.  All `override` properties are of type `String`.
```
Comment (keiramoon) [inline Packages/com.unity.inputsystem/Documentation~/ActionBindings.md:374]:
```suggestion
You can override aspects of any Binding at runtime non-destructively. Specific properties of [`InputBinding`](../api/UnityEngine.InputSystem.InputBinding.html) have an `override` variant that, if set, will take precedent over the property that they shadow.  All `override` properties are of type `String`.
```

File: Packages/com.unity.inputsystem/Documentation~/ActionBindings.md [hunk 12]
```diff
@@ -261,58 +286,162 @@ public class CustomParameterEditor : InputParameterEditor<CustomComposite>
...
+
+In most cases, it is best to locate specific bindings using APIs such as [`GetBindingIndexForControl`](../api/UnityEngine.InputSystem.InputActionRebindingExtensions.html#UnityEngine_InputSystem_InputActionRebindingExtensions_GetBindingIndexForControl_) and to then apply the override to that specific binding.
 
->NOTE: To download a sample project which demonstrates how to set up a rebinding user interface with Input System APIs, open the Package Manager, select the Input System Package, and choose the sample project "Rebinding UI" to download.
+```CSharp
+// Find the "Jump" binding for the space key.
+var jumpAction = playerInput.actions["Jump"];
+var bindingIndex = jumpAction.GetBindingIndexForControl(Keyboard.current.spaceKey);
+
+// And change it to the enter key.
```
Comment (keiramoon) [inline Packages/com.unity.inputsystem/Documentation~/ActionBindings.md:398]:
```suggestion
// And change it to the Enter key.
```

File: Packages/com.unity.inputsystem/Documentation~/ActionBindings.md [hunk 13]
```diff
@@ -261,58 +286,162 @@ public class CustomParameterEditor : InputParameterEditor<CustomComposite>
...
+// Can also do this by using the name given to the composite binding.
+playerInput.actions["move"].ChangeCompositeBinding("WASD").Erase();
+
+// Erase first binding in "gameplay" action map.
+playerInput.actions.FindActionMap("gameplay").ChangeBinding(0).Erase();
+```
+
+### Adding Bindings
+
+New bindings can be added to an Action using [`AddAction`](../api/UnityEngine.InputSystem.InputActionSetupExtensions.html#UnityEngine_InputSystem_InputActionSetupExtensions_AddBinding_UnityEngine_InputSystem_InputAction_System_String_System_String_System_String_System_String_) or [`AddCompositeBinding`](../api/UnityEngine.InputSystem.InputActionSetupExtensions.html#UnityEngine_InputSystem_InputActionSetupExtensions_AddCompositeBinding_UnityEngine_InputSystem_InputAction_System_String_System_String_System_String_).
```
Comment (keiramoon) [inline Packages/com.unity.inputsystem/Documentation~/ActionBindings.md:423]:
```suggestion
To add new bindings to an Action, use [`AddAction`](../api/UnityEngine.InputSystem.InputActionSetupExtensions.html#UnityEngine_InputSystem_InputActionSetupExtensions_AddBinding_UnityEngine_InputSystem_InputAction_System_String_System_String_System_String_System_String_) or [`AddCompositeBinding`](../api/UnityEngine.InputSystem.InputActionSetupExtensions.html#UnityEngine_InputSystem_InputActionSetupExtensions_AddCompositeBinding_UnityEngine_InputSystem_InputAction_System_String_System_String_System_String_).
```

File: Packages/com.unity.inputsystem/Documentation~/Testing.md (hunk 2 above)
Comment (karljj1) [reply Packages/com.unity.inputsystem/Documentation~/Testing.md:10]: Try using lowercase json. Its the first example https://www.markdownguide.org/extended-syntax ;)

Comment (karljj1) [reply Packages/com.unity.inputsystem/Documentation~/Testing.md:10]: Although the old code works for me in my md viewer so maybe our generator is bugged.

File: Packages/com.unity.inputsystem/Documentation~/ActionBindings.md [hunk 14]
```diff
@@ -338,7 +467,38 @@ Refer to the scripting API reference for [`InputActionRebindingExtensions.Rebind
...
+// Restore player rebinds from PlayerPrefs (removes all existing
+// overrides on the actions; pass `false` for second argument
+// in case you want to prevent that).
+var rebinds = PlayerPrefs.GetString("rebinds");
+playerInput.actions.LoadBindingOverridesFromJson(rebinds);
+```
+
+### Restoring original Bindings
+
+You can remove Binding overrides and thus restore defaults by using [`RemoveBindingOverride`](../api/UnityEngine.InputSystem.InputActionRebindingExtensions.html#UnityEngine_InputSystem_InputActionRebindingExtensions_RemoveBindingOverride_UnityEngine_InputSystem_InputAction_System_Int32_) or [`RemoveAllBindingOverrides`](../api/UnityEngine.InputSystem.InputActionRebindingExtensions.html#UnityEngine_InputSystem_InputActionRebindingExtensions_RemoveAllBindingOverrides_UnityEngine_InputSystem_IInputActionCollection2_).
```
Comment (keiramoon) [inline Packages/com.unity.inputsystem/Documentation~/ActionBindings.md:488]:
```suggestion
To remove Binding overrides and thus restore defaults, use [`RemoveBindingOverride`](../api/UnityEngine.InputSystem.InputActionRebindingExtensions.html#UnityEngine_InputSystem_InputActionRebindingExtensions_RemoveBindingOverride_UnityEngine_InputSystem_InputAction_System_Int32_) or [`RemoveAllBindingOverrides`](../api/UnityEngine.InputSystem.InputActionRebindingExtensions.html#UnityEngine_InputSystem_InputActionRebindingExtensions_RemoveAllBindingOverrides_UnityEngine_InputSystem_IInputActionCollection2_).
```

File: Packages/com.unity.inputsystem/Documentation~/ActionBindings.md [hunk 15]
```diff
@@ -399,12 +559,59 @@ You can also use this method to replace the text string with images.
...
+
+* A Binding path can also contain wildcards, such as `<Gamepad>/button*`. This matches any Control on any gamepad with a name starting with "button", which matches all the four action buttons on any connected gamepad. A different example: `*/{Submit}` matches any Control tagged with the "Submit" [usage](Controls.md#control-usages) on any Device.
+
+To query the Controls that an Action resolves to, you can use [`InputAction.controls`](../api/UnityEngine.InputSystem.InputAction.html#UnityEngine_InputSystem_InputAction_controls). You can also run this query if the Action is disabled.
+
+#### Choosing which Devices to use
+
+>__Note__: [`InputUser`](UserManagement.md) and [`PlayerInput`](Components.md) make use of this facility automatically. They set [`InputActionMap.devices`](../api/UnityEngine.InputSystem.InputActionMap.html#UnityEngine_InputSystem_InputActionMap_devices) automatically based on the Devices that are paired to the user.
```
Comment (keiramoon) [inline Packages/com.unity.inputsystem/Documentation~/ActionBindings.md:590]: Can we consolidate this Note like this?

```suggestion
>__Note__: [`InputUser`](UserManagement.md) and [`PlayerInput`](Components.md) automatically set [`InputActionMap.devices`](../api/UnityEngine.InputSystem.InputActionMap.html#UnityEngine_InputSystem_InputActionMap_devices) based on the Devices that are paired to the user.
```

File: Packages/com.unity.inputsystem/Documentation~/ActionBindings.md [hunk 16]
```diff
@@ -399,12 +559,59 @@ You can also use this method to replace the text string with images.
...
+
+    // Restrict the action map to just the first gamepad.
+    actionMap.devices = new[] { Gamepad.all[0] };
+```
+
+### Disambiguation
+
+If multiple Controls are bound to an Action, the Input System monitors input from each bound Control to feed the Action. The Input System must also define which of the bound controls to use for the value of the action. For example, if you have a Binding to `<Gamepad>/leftStick`, and you have multiple connected gamepads, the Input System must determine which gamepad's stick provides the input value for the Action.
+
+This Control is the "driving" Control; the Control which is driving the Action. Unity decides which Control is currently driving the Action in a process called disambiguation.
```
Comment (keiramoon) [inline Packages/com.unity.inputsystem/Documentation~/ActionBindings.md:607]:
```suggestion
This Control is the "driving" Control: the Control which is driving the Action. Unity decides which Control is currently driving the Action in a process called disambiguation.
```

File: Packages/com.unity.inputsystem/Documentation~/Gamepad.md [hunk 17]
```diff
@@ -6,6 +6,8 @@ A gamepad can have additional Controls, such as a gyro, which the Device can exp
 
 Gamepad support guarantees the correct location and functioning of Controls across platforms and hardware. For example, a PS4 DualShock controller layout should look identical regardless of which platform it is supported on. A gamepad's south face button should always be the lowermost face button.
 
+>NOTE: Generic [HID](./HID.md) gamepads will __not__ be surfaced as [`Gamepad`](../api/UnityEngine.InputSystem.Gamepad.html) devices but rather be created as generic [joysticks](./Joystick.md). This is because the Input System cannot guarantee correct mapping of buttons and axes on the controller (the information is simply not available at the HID level). Only HID gamepads that are explicitly supported by the Input System (like the PS4 controller) will come out as gamepads. Note that you can set up the same kind of support for specific HID gamepads yourself (see ["Overriding the HID Fallback"](./HID.md#overriding-the-hid-fallback)).
```
Comment (keiramoon) [inline Packages/com.unity.inputsystem/Documentation~/Gamepad.md:9]:
```suggestion
>NOTE: Generic [HID](./HID.md) gamepads will __not__ be surfaced as [`Gamepad`](../api/UnityEngine.InputSystem.Gamepad.html) devices, but rather be created as generic [joysticks](./Joystick.md). This is because the Input System cannot guarantee correct mapping of buttons and axes on the controller (the information is not available at the HID level). Only HID gamepads that are explicitly supported by the Input System (like the PS4 controller) will come out as gamepads. You can set up the same kind of support for specific HID gamepads yourself (see ["Overriding the HID Fallback"](./HID.md#overriding-the-hid-fallback)).
```

File: Packages/com.unity.inputsystem/Documentation~/KnownLimitations.md [hunk 18]
```diff
@@ -33,3 +33,4 @@ The following is a list of known limitations that the Input System currently has
 ## Features Supported by Old Input Manager
 
 * `MonoBehaviour` mouse methods (`OnMouseEnter`, `OnMouseDrag`, etc) will not be called by the Input System.
+* The Unity Remote does not yet have support for the Input System. This is being worked on.
```
Comment (keiramoon) [inline Packages/com.unity.inputsystem/Documentation~/KnownLimitations.md]:
```suggestion
* Unity Remote doesn't currently support the Input System. This is being worked on.
```

File: Packages/com.unity.inputsystem/Documentation~/OnScreen.md [hunk 19]
```diff
@@ -18,17 +18,31 @@ To query the Control (and, implicitly, the Device) that an on-screen Control fee
 
 ## On-screen buttons
 
-The [`OnScreenButton`](../api/UnityEngine.InputSystem.OnScreen.OnScreenButton.html) component requires the target Control to be a `Button` Control. [`OnScreenButton`](../api/UnityEngine.InputSystem.OnScreen.OnScreenButton.html) sets the target Control value to 1 when it receives a pointer-down (`IPointerDownHandler.OnPointerDown`) event, or 0 when it receives a pointer-up (`IPointerUpHandler.OnPointerUp`) event.
+To create an on-screen button:
+
+1. Add a UI `Button` object.
+2. Add the [`OnScreenButton`](../api/UnityEngine.InputSystem.OnScreen.OnScreenButton.html) component to it.
+3. Set the [`Control Path`](../api/UnityEngine.InputSystem.OnScreen.OnScreenControl.html#UnityEngine_InputSystem_OnScreen_OnScreenControl_controlPath) to refer to a [`ButtonControl`](../api/UnityEngine.InputSystem.Controls.ButtonControl.html). For example, `<Gamepad>/buttonSouth`. The type of device referenced by the control path determines the type of virtual device created by the component.
```
Comment (keiramoon) [inline Packages/com.unity.inputsystem/Documentation~/OnScreen.md]:
```suggestion
3. Set the [`Control Path`](../api/UnityEngine.InputSystem.OnScreen.OnScreenControl.html#UnityEngine_InputSystem_OnScreen_OnScreenControl_controlPath) to refer to a [`ButtonControl`](../api/UnityEngine.InputSystem.Controls.ButtonControl.html) (for example, `<Gamepad>/buttonSouth`). The type of device referenced by the control path determines the type of virtual device created by the component.
```

File: Packages/com.unity.inputsystem/Documentation~/OnScreen.md [hunk 20]
```diff
@@ -18,17 +18,31 @@ To query the Control (and, implicitly, the Device) that an on-screen Control fee
...
 
+The [`OnScreenButton`](../api/UnityEngine.InputSystem.OnScreen.OnScreenButton.html) component requires the target Control to be a `Button` Control. [`OnScreenButton`](../api/UnityEngine.InputSystem.OnScreen.OnScreenButton.html) sets the target Control value to 1 when it receives a pointer-down (`IPointerDownHandler.OnPointerDown`) event, or 0 when it receives a pointer-up (`IPointerUpHandler.OnPointerUp`) event.
+
 ## On-screen sticks
 
+To create an on-screen stick:
+
+1. Create a UI `Image` object.
+2. Add the [`OnScreenStick`](../api/UnityEngine.InputSystem.OnScreen.OnScreenStick.html) component to it.
+3. Set the [`Control Path`](../api/UnityEngine.InputSystem.OnScreen.OnScreenControl.html#UnityEngine_InputSystem_OnScreen_OnScreenControl_controlPath) to refer to a [`Vector2Control`](../api/UnityEngine.InputSystem.Controls.Vector2Control.html). For example, `<Gamepad>/leftStick`. The type of device referenced by the control path determines the type of virtual device created by the component.
```
Comment (keiramoon) [inline Packages/com.unity.inputsystem/Documentation~/OnScreen.md]:
```suggestion
3. Set the [`Control Path`](../api/UnityEngine.InputSystem.OnScreen.OnScreenControl.html#UnityEngine_InputSystem_OnScreen_OnScreenControl_controlPath) to refer to a [`Vector2Control`](../api/UnityEngine.InputSystem.Controls.Vector2Control.html) (for example, `<Gamepad>/leftStick`). The type of device referenced by the control path determines the type of virtual device created by the component.
```

File: Packages/com.unity.inputsystem/Documentation~/Processors.md [hunk 21]
```diff
@@ -27,6 +27,24 @@ An Input Processor takes a value and returns a processed result for it. The rece
 
 You can install Processors on [bindings](ActionBindings.md), [actions](Actions.md) or on [controls](Controls.md).
 
+Each Processor is [registered](../api/UnityEngine.InputSystem.InputSystem.html#UnityEngine_InputSystem_InputSystem_RegisterProcessor__1_System_String_) using a unique name. You can replace an existing Processor by registering your own Processor under an existing name.
```
Comment (keiramoon) [inline Packages/com.unity.inputsystem/Documentation~/Processors.md]:
```suggestion
Each Processor is [registered](../api/UnityEngine.InputSystem.InputSystem.html#UnityEngine_InputSystem_InputSystem_RegisterProcessor__1_System_String_) using a unique name. To replace an existing Processor, register your own Processor under an existing name.
```

File: Packages/com.unity.inputsystem/Documentation~/Processors.md [hunk 22]
```diff
@@ -209,6 +227,8 @@ A stick deadzone Processor scales the values of a Vector2 Control, such as a sti
 
 You can also write custom Processors to use in your Project. Custom Processors are available in the UI and code in the same way as the built-in Processors. Add a class derived from [`InputProcessor<TValue>`](../api/UnityEngine.InputSystem.InputProcessor-1.html), and implement the [`Process`](../api/UnityEngine.InputSystem.InputProcessor-1.html#UnityEngine_InputSystem_InputProcessor_1_Process__0_UnityEngine_InputSystem_InputControl_) method:
 
+>__IMPORTANT__: Processors must be __stateless__. This means you cannot store local state in a processor that will change depending on the input being processed. The reason for this is because processors are not part of the [input state](./Controls.md#control-state) being kept by the input system.
```
Comment (keiramoon) [inline Packages/com.unity.inputsystem/Documentation~/Processors.md]:
```suggestion
>__IMPORTANT__: Processors must be __stateless__. This means you cannot store local state in a processor that will change depending on the input being processed. The reason for this is because processors are not part of the [input state](./Controls.md#control-state) that the Input System keeps.
```

**Review by keiramoon — APPROVED** [review]
Approved (no ship-stoppers) but please see comments and suggestions :)

File: CONTRIBUTIONS.md (hunk 6 above)
Comment (Rene-Damm) [reply CONTRIBUTIONS.md]: I think that part does matter. Making a pull request, I'd expect someone to either take it or reject it. Whereas in our current workflow, even if we do take it, it will end up getting closed and not merged and your contribution won't really show as such in the history.

File: CONTRIBUTIONS.md (hunk 4 above)
Comment (Rene-Damm) [reply CONTRIBUTIONS.md]: Pushed

File: CONTRIBUTIONS.md (hunk 5 above)
Comment (Rene-Damm) [reply CONTRIBUTIONS.md]: Pushed

File: CONTRIBUTIONS.md (hunk 6 above)
Comment (Rene-Damm) [reply CONTRIBUTIONS.md]: Pushed

File: CONTRIBUTIONS.md (hunk 8 above)
Comment (Rene-Damm) [reply CONTRIBUTIONS.md]: Pushed

File: CONTRIBUTIONS.md (hunk 7 above)
Comment (Rene-Damm) [reply CONTRIBUTIONS.md]: Pushed

File: Packages/com.unity.inputsystem/Documentation~/OnScreen.md (hunk 19 above)
Comment (Rene-Damm) [reply Packages/com.unity.inputsystem/Documentation~/OnScreen.md]: Pushed

File: Packages/com.unity.inputsystem/Documentation~/Processors.md (hunk 21 above)
Comment (Rene-Damm) [reply Packages/com.unity.inputsystem/Documentation~/Processors.md]: Pushed

File: Packages/com.unity.inputsystem/Documentation~/Processors.md (hunk 22 above)
Comment (Rene-Damm) [reply Packages/com.unity.inputsystem/Documentation~/Processors.md]: Pushed

File: Packages/com.unity.inputsystem/Documentation~/OnScreen.md (hunk 20 above)
Comment (Rene-Damm) [reply Packages/com.unity.inputsystem/Documentation~/OnScreen.md]: Pushed

File: Packages/com.unity.inputsystem/Documentation~/KnownLimitations.md (hunk 18 above)
Comment (Rene-Damm) [reply Packages/com.unity.inputsystem/Documentation~/KnownLimitations.md]: Pushed
